flang/lib/Lower/Mangler.cpp

```diff
@@ -99,11 +99,14 @@ Fortran::lower::mangle::mangleName(const Fortran::semantics::Symbol &symbol,
 
   // TODO: A procedure that inherits BIND(C) through another interface
   // (procedure(iface)) should be dealt with in GetBindName() or some wrapper.
-  if (!Fortran::semantics::IsPointer(ultimateSymbol) &&
-      Fortran::semantics::IsBindCProcedure(ultimateSymbol) &&
-      Fortran::semantics::ClassifyProcedure(symbol) !=
-          Fortran::semantics::ProcedureDefinitionClass::Internal)
-    return ultimateSymbol.name().ToString();
+  if (const auto *procDetails{
```

Comment:
@klausler, the code here is to deal correctly with the case:
```
 interface
   subroutine iface() bind(c)
   end subroutine
 end interface
 procedure(iface) :: foo
 call foo()
end
```
foo has a binding label, but GetBindName() returns a nullptr. Ideally it should return a pointer to "foo".

I could not come to a solid confidence that we always have enough info already in name resolution to set this in SetBindNameOn, so I kept the special handling here in lowering where I am sure `procDetails->procInterface()` was set.

Reply:
Why would `foo` have a binding label?

Reply:
> We should probably check that a BIND(C,NAME='...') suffix on an external procedure interface declaration (with an explicit name) doesn't exist when the interface is used in a procedure-declaration-stmt that isn't a pointer -- there's nothing in the standard about that case, but it would lead to trouble.

Do you mean that the following should be forbidden?
```
 interface
   subroutine iface() bind(c, "some_c_name")
   end subroutine
 end interface
 procedure(iface) :: foo
 call foo()
end
```

> You could call SetBindNameOn(symbol) on foo as part of procedure-declaration-stmt name resolution

So you think we can already call `Fortran::semantics::IsBindCProcedure(symbol)`instead of using `attrs_->test(Attr::BIND_C)` in SetBindNameOn?

Reply:
> Do you mean that the following should be forbidden?
> 
> ```
>  interface
>    subroutine iface() bind(c, "some_c_name")
>    end subroutine
>  end interface
>  procedure(iface) :: foo
>  call foo()
> end
> ```

What would it mean?  Does it make 'foo' an alias for 'iface', using the same binding label?  Does it mean that `foo` has the `BIND(C)` attribute with no explicit binding label?  The standard doesn't say.

>  > You could call SetBindNameOn(symbol) on foo as part of procedure-declaration-stmt name resolution
> 
> So you think we can already call `Fortran::semantics::IsBindCProcedure(symbol)`instead of using `attrs_->test(Attr::BIND_C)` in SetBindNameOn?

Why would you want to make that change?

Also - a procedure-declaration-stmt *does* allow a `BIND(C[,NAME=...])` attribute -- I didn't notice it in the syntax because it was indirected.  So now we have to worry about the interpretation of having `BIND(C)` on the explicit interface declaration and on the procedure-declaration-stmt.

Reply:
> What would it mean? Does it make 'foo' an alias for 'iface', using the same binding label? Does it mean that foo has the BIND(C) attribute with no explicit binding label? The standard doesn't say.

I agree the standard is not very specific about these cases.... Now, the "normal" symbol name is part of the interface, yet a procedure statement does not create an alias of the interface name, so I would follow the same logic and say that the NAME="" in an interface does not become the binding label of all the procedure using that interface in a procedure statement. I would say that BIND(C), BIND(C, name="some_name") and BIND(C, name="") on an interface all have the same effect on procedures in procedure statement: they are as if the procedure was BIND(C).
 
This matches what nag, gfortran and ifort do (I cannot completely tell for XLF since it does not append underscore), nvfortran does not implicitly give a binding label to procedures in procedure statement with a BIND(C) iface. So all compilers accept NAME= on an interface and do not propagate the _explicit_ biding label on the procedures using that interface in procedure statements.

Here is a test case:

```
subroutine test() 
 interface 
   subroutine notbindcatall() 
   end subroutine 
   subroutine iface_bindc() bind(c) 
   end subroutine 
   subroutine iface_explicit_name() bind(c, name="explicit_name") 
   end subroutine 
   subroutine iface_nobinding() bind(c, name="") 
   end subroutine 
 end interface 
 
 procedure(iface_bindc) :: foo_iface_bindc 
 procedure(iface_explicit_name) :: foo_iface_explicit_name 
 procedure(iface_nobinding) :: foo_iface_nobinding 
  
 procedure(iface_bindc),  bind(c, name="bar_iface_bindc_2") :: bar_iface_bindc 
 procedure(iface_explicit_name),  bind(c,name="bar_iface_explicit_name_2") ::  bar_iface_explicit_name 
 procedure(iface_nobinding), bind(c, name="bar_iface_nobinding_2") :: bar_iface_nobinding 
  
 procedure(iface_bindc),  bind(c, name="") :: nobinding_iface_bindc 
 procedure(iface_explicit_name),  bind(c, name="") :: nobinding_iface_explicit_name 
 procedure(iface_nobinding), bind(c, name="") :: nobinding_iface_nobinding 
 
 call notbindcatall() 
 
 call iface_bindc() 
 call iface_explicit_name() 
 call iface_nobinding() 
 
 call foo_iface_bindc() 
 call foo_iface_explicit_name() 
 call foo_iface_nobinding() 
 
 call bar_iface_bindc() 
 call bar_iface_explicit_name() 
 call bar_iface_nobinding() 
 
 call nobinding_iface_bindc() 
 call nobinding_iface_explicit_name() 
 call nobinding_iface_nobinding() 
end subroutine
```

gfortran and nag names looks like this:
```
        call    notbindcatall_

        call    iface_bindc
        call    explicit_name
        call    iface_nobinding_

        call    foo_iface_bindc
        call    foo_iface_explicit_name
        call    foo_iface_nobinding

        call    bar_iface_bindc_2
        call    bar_iface_explicit_name_2
        call    bar_iface_nobinding_2

        call    nobinding_iface_bindc_
        call    nobinding_iface_explicit_name_
        call    nobinding_iface_nobinding_
```

ifort does not enforce the fact that NAME="" drops the binding label of iface_nobinding_ and nobinding_xxx but is otherwise similar.

Reply:
@klausler, I updated the patch to resolve the binding label in semantics when the interface is BIND(C) as described above.

Reply:
Thank you.  Before I forget, did you make the name mangling target-specific?  I seem to recall that XLF does not append an underscore character.

Reply:
> Before I forget, did you make the name mangling target-specific?

In this patch it does not matter, the test is added at the FIR level, so non BIND(C) procedure IR name use FIR internal mangling that is target independent (start with _QP...., later transformed to target dependent names for external procedures when producing LLVM from ). The BIND(C) IR name are just the binding labels in FIR and LLVM IR.

But in previous BIND(C) patch to catch assembly conflict, yes, I had to deal with the underscoring aspect by propagating the option from the driver to semantic [here](https://github.com/llvm/llvm-project/blob/1062c140f811fabb7a926fde74112ea4d2330b74/flang/lib/Semantics/check-declarations.cpp#L2641).

> I seem to recall that XLF does not append an underscore character.

That is right, flang-new/gfortran can also do that with the -fno-undescoring option.

Reply:
@klausler, are you ok with the patch?